### DISCUSSION TITLE
Custom Stream impl not getting polled

### DISCUSSION THREAD

**Discussion by cdmistman** [tokio-rs/tokio#2689] [General] [answered]
**Version**
tokio v0.2.22
tokio-macros v0.2.5

**Platform**
Linux 5.4.0-7634-generic x86_64

**Description**
I've written a single-producer single-consumer structure that implements `tokio::stream::Stream` on the consumer end. The point of the structure is to both order values and to release each value on a timer (with a potential "wakeup" delay). After adding prints, it appears that the producer is successfully adding the values, but the consumer stops getting polled after being awaited.

The repro is [here](https://github.com/cdmistman/tokio_buffers). To see it, `cargo test -- --nocapture` will show that the producer is inserting values, but the consumer is not printing the `Poll` value, indicating that it is not being polled. Changing the number of values sent in the test to a much more manageable number (eg 50) results in all of the values being sent, but none being consumed, confirming that the `OrderedTimedReleaseConsumer` is not being polled.

**Comment by Darksonn** [comment, marked as answer]
If you return `Poll::Pending`, then that is a promise that you will send a wake-up notification to the waker in the provided context once you are ready to continue. It is not polled because you are not sending any wake-ups.

You likely want to do this by polling a `Delay`, but you should note that just creating it is not enough. You have to actually poll it. Check out our [async in depth][1] guide for more information.

[1]: https://tokio.rs/tokio/tutorial/async

**Comment by cdmistman** [comment]
This worked, thanks!
